crates/bevy_pbr/src/material_bind_groups.rs

```diff
@@ -155,634 +276,1150 @@ impl From<u32> for MaterialBindGroupIndex {
+            resource_to_slot: HashMap::default(),
+            free_slots: vec![],
+            len: 0,
         }
     }
 
-    /// Allocates a new slot and returns its index.
+    /// Returns the slot corresponding to the given resource, if that resource
+    /// is located in this binding array.
     ///
-    /// This bind group must not be full.
-    fn allocate(&mut self) -> MaterialBindGroupSlot {
-        debug_assert!(!self.is_full());
-        self.allocation = MaterialNonBindlessBindGroupAllocation::Allocated;
-        MaterialBindGroupSlot(0)
+    /// If the resource isn't in this binding array, this method returns `None`.
+    fn find(&self, binding_resource_id: BindingResourceId) -> Option<u32> {
+        self.resource_to_slot.get(&binding_resource_id).copied()
     }
 
-    /// Assigns an unprepared bind group to the group and slot specified in the
-    /// [`MaterialBindingId`].
-    ///
-    /// For non-bindless bind groups, we go ahead and create the bind group
-    /// immediately.
-    fn init(
-        &mut self,
-        render_device: &RenderDevice,
-        bind_group_layout: &BindGroupLayout,
-        _: MaterialBindGroupSlot,
-        unprepared_bind_group: UnpreparedBindGroup<M::Data>,
-    ) {
-        let entries = unprepared_bind_group
-            .bindings
-            .iter()
-            .map(|(index, binding)| BindGroupEntry {
-                binding: *index,
-                resource: binding.get_binding(),
-            })
-            .collect::<Vec<_>>();
-
-        self.allocation = MaterialNonBindlessBindGroupAllocation::Initialized(
-            render_device.create_bind_group(M::label(), bind_group_layout, &entries),
-            unprepared_bind_group.data,
-        );
-    }
-
-    /// Fills the slot directly with a custom bind group.
-    ///
-    /// This is only a meaningful operation for non-bindless bind groups. It's
-    /// rarely used, but see the `texture_binding_array` example for an example
-    /// demonstrating how this feature might see use in practice.
-    fn init_custom(&mut self, bind_group: BindGroup, extra_data: M::Data) {
-        self.allocation =
-            MaterialNonBindlessBindGroupAllocation::Initialized(bind_group, extra_data);
-    }
+    /// Inserts a bindless resource into a binding array and returns the index
+    /// of the slot it was inserted into.
+    fn insert(&mut self, binding_resource_id: BindingResourceId, resource: R) -> u32 {
+        let slot = self.free_slots.pop().unwrap_or(self.len);
+        self.resource_to_slot.insert(binding_resource_id, slot);
 
-    /// Deletes the stored bind group.
-    fn free(&mut self, _: MaterialBindGroupSlot) {
-        self.allocation = MaterialNonBindlessBindGroupAllocation::Unallocated;
-    }
+        if self.bindings.len() < slot as usize + 1 {
+            self.bindings.resize_with(slot as usize + 1, || None);
+        }
+        self.bindings[slot as usize] = Some(MaterialBindlessBinding::new(resource));
```

Comment:
Why not push?

Reply:
I think it's slightly less code this way. Otherwise it'd be something like:
```rust
if self.bindings.len() < slot as usize + 1 {
    if self.bindings.len() < slot as usize {
        self.bindings.resize_with(slot as usize, || None);
    }
    self.bindings.push(Some(MaterialBindlessBinding::new(resource));
} else {
    self.bindings[slot as usize] = Some(MaterialBindlessBinding::new(resource));
}
```